Feat/handle error transform

tiagralee opened the pull request:
issue #1 
- Add a common.CustomError
- Changed `svc_handler.sysl` to transform the error type to the CustomerError map
- Changed `svc_handler.sysl` add a new function HandleError to map custom error, if not found return the common http error.
- Update the tests for simple and dbendpoints.

gkanz commented:
@tiagralee, could you please add a summary to the PR regarding why you chose to approach you did to meet the requirement given in #1? There are no commit comments for me to go on...

tiagralee commented:
@gkanz Thanks Greg, just add a summary.

andrewemeryanz commented on codegen/tests/dbendpoints/dbendpoints_gencallback.go:
```diff
@@ -30,12 +30,10 @@ func (c Callback) Config() validator.Validator {
 }
 
-func (c Callback) HandleError(ctx context.Context, w http.ResponseWriter, kind common.Kind, message string, cause error) {
+func (c Callback) MapError(ctx context.Context, w http.ResponseWriter, kind common.Kind, message string, cause error) common.HTTPError {
```
`MapError` shouldn't get access to a `ResponseWriter`.

andrewemeryanz commented on codegen/tests/dbendpoints/servicehandler.go:
```diff
@@ -34,10 +35,16 @@ func NewServiceHandler(genCallback GenCallback, serviceInterface *ServiceInterfa
 }
 
+// Handler Error
+func (s *ServiceHandler) HandleError(ctx context.Context, w http.ResponseWriter, kind common.Kind, message string, cause error) {
```
`HandleError` can be private: `handleError`

tiagralee replied:
Thanks Andrew, just make this lowercase will make it private?

andrewemeryanz replied:
Yes.

andrewemeryanz reviewed (changes requested):
Minor changes requested.

gkanz commented:
I have some questions about the approach: I've added a comment to https://github.com/anz-bank/sysl-go/issues/1 with my thoughts.

tiagralee commented:
Thanks Greg, I'll have a look.

tiagralee commented on codegen/tests/simple/servicehandler.go:
```diff
@@ -35,10 +42,32 @@ func NewServiceHandler(genCallback GenCallback, serviceInterface *ServiceInterfa
 	return &ServiceHandler{genCallback, serviceInterface}
 }
 
```
As the current transformation does not support switch statement, use ifelse statement instead, the below code equivalent to `
serverError := common.CreateError(ctx, kind, message, cause)
	httpError := s.genCallback.MapError(ctx, serverError)
	if httpError == nil {
		switch t := cause.(type) {
                 case CustomError:
                      e := t.HTTPError()
                      httpError = &e
                 default:
                      e := common.HandleError(ctx, serverError)
                      httpError = &e
                }
	}
	httpError.WriteError(ctx, w)
`

andrewemeryanz commented on codegen/tests/simple/servicehandler.go:
```diff
@@ -11,6 +12,12 @@ import (
+var businesslogicerror = map[string]string{"name": "BusinessLogicError", "http_code": "1001", "http_message": "foo", "http_status": "500"}
+
+// *BusinessLogicError2 error
+var businesslogicerror2 = map[string]string{"name": "BusinessLogicError2", "http_code": "1002", "http_message": "foo2", "http_status": "501"}
```
`businesslogicerror2` needs to be capitalised to be public: 
```go
var BusinessLogicError2 = ...
```

Please make this value be the name of the type with a capital letter at the start:
```text
!type FooBar [~error]: ...     # becomes: val FooBar = ...
!type foobar [~error]: ...     # becomes: val Foobar = ...
```

andrewemeryanz commented on codegen/tests/simple/simple_gencallback.go:
```diff
@@ -27,12 +26,9 @@ func (c Callback) Config() validator.Validator {
-
-	httpError.WriteError(ctx, w)
+func (c Callback) MapError(ctx context.Context, cause error) *common.HTTPError {
+	handleError := common.HandleError(ctx, cause)
```
You can just return `nil` here. `common.HandlerError` is called by the `handlerError` function internally.

andrewemeryanz commented on codegen/tests/simple/types.go:
```diff
@@ -16,6 +16,13 @@ var _ = time.Parse
 var _ = date.Parse
 
+// BusinessLogicError ...
+type BusinessLogicError struct {
```
We shouldn't be generating types here for items tagged with `[~error]`.

andrewemeryanz commented on common/custom_error.go:
```diff
@@ -0,0 +1,30 @@
+func (e CustomError) HTTPError() HTTPError {
+	httpStatus, err := strconv.Atoi(getOrDefault(e, "http_status", "500"))
+	if err != nil {
+		httpStatus = 500
```
We should be logging this error. Pass in the `Context` if necessary.

tiagralee replied:
The only reference is from the HandleError, which log the error before calling this.

andrewemeryanz replied:
Sorry, I mean the `err` returned as a result of `strconv.Atoi`.

andrewemeryanz replied:
I was thinking something more like:
```
func (e CustomError) HTTPError(ctx context.Context) HTTPError {
	httpStatusString := getOrDefault(e, "http_status", "500")
	httpStatus, err := strconv.Atoi(httpStatusString)
	if err != nil {
		logEntry := GetLogEntryFromContext(ctx)
		logEntry.Error(fmt.Sprintf("invalid http_status: %s for error: %s, returning 500", httpStatusString, e["name"]))
		httpStatus = 500
	}
```

Now `500` should really be a constant somewhere in `sysl-go` but let's park that for now. Feel free to put a `TODO` in there for now.

andrewemeryanz commented on codegen/tests/simple/servicehandler.go:
```diff
@@ -35,10 +42,32 @@ func NewServiceHandler(genCallback GenCallback, serviceInterface *ServiceInterfa
 }
 
+// Handler Error
+func (s *ServiceHandler) handleError(ctx context.Context, w http.ResponseWriter, kind common.Kind, message string, cause error) {
```
This function doesn't need to be generated because it doesn't contain any custom code. The following code could replace `common.HandleError`:

```go
func HandleError(ctx context.Context, w http.ResponseWriter, kind Kind, message string, cause error,
		httpErrorMapper func(context.Context, ServerError) *HTTPError) {

	logEntry := GetLogEntryFromContext(ctx)
	logEntry.Error(cause)

	err := CreateError(ctx, kind, message, cause)
	httpError := httpErrorMapper(ctx, err)
	if httpError == nil {
		switch t := cause.(type) {
		case CustomError:
			e := t.HTTPError()
			httpError = &e
		default:
			e := mapError(ctx, &err)
			httpError = &e
		}
	}
	httpError.WriteError(ctx, w)
}

func mapError(ctx context.Context, err error) HTTPError {

	var (
		httpCode        int
		errorCode, desc string
	)
```

andrewemeryanz reviewed (changes requested):
We're getting closer. We can avoid a bunch of generated code by doing more in `sysl-go`.

tiagralee commented on common/error_handler.go:
```diff
@@ -13,7 +14,27 @@ const (
+	httpError.WriteError(ctx, w)
+}
+
+func MapError(ctx context.Context, err error) HTTPError {
```
Need to make the MapError public, as the defaultcallback need to use this.

andrewemeryanz replied:
> Need to make the MapError public

Yes. That's fine.

andrewemeryanz reviewed (changes requested):
One minor change, then we'll tidy up the commit history to make it more readable.

Commit:
Move error handler to error mapper

This change create a common.CustomError and update the svc.handler.sysl use cusom HTTPError to map the error and wirtes the error
